### PR TITLE
Deterministic cudnn algorithms

### PR DESCRIPTION
**Pull request by prigoyal**
addresses #114

### CODE REVIEW

File: torch/csrc/cudnn/Conv.cpp [hunk 1]
```diff
@@ -220,68 +303,119 @@ struct algorithm_search<cudnnConvolutionBwdDataAlgo_t> {
...
+  // declare an instance of the algo class {fwd, bwd_filter, bwd_data} that we
+  // want to search on
   using search = algorithm_search<algo_t>;
   auto& cache = search::cache();
 
   if (cache.find(conv.params, algo)) {
     return;
   }
 
+  if (deterministic) {
```
Comment (prigoyal) [inline torch/csrc/cudnn/Conv.cpp]: this is the relevant change in this file for determinism however I found the code slightly hard to read so I made some style changes. If people don't like them I can revert as well but I find this code much more readable now

**Comment by ezyang** [comment]
LGTM. For future reference, it makes it easier to review if style changes are placed in a separate commit. In this case it was not hard to follow and I wouldn't bother going back and fixing it.

**Comment by soumith** [comment]
I'm pretty sure this PR doesn't work and will have runtime failures (even if you run unit tests).

I dont see how the torch.backends.cudnn.deterministic flag actually propagates into Convolution.cpp
See how the cudnn.benchmark flag propagates through, for example see https://github.com/pytorch/pytorch/blob/9be8d0a9d2492732218a497a0fff2b41255c82c5/torch/nn/functional.py

**Comment by ngimel** [comment]
Starting from cudnn v6 the structure returned by cudnnFindAlgorithm contains determinism field. From performance point of view, if benchmarking is on, and cudnnFind is called anyway, it is better to return the whole structure (rather than only the first member) and pick the fastest deterministic algo.

File: torch/csrc/cudnn/Conv.cpp [hunk 2]
```diff
@@ -163,53 +203,96 @@ struct algorithm_search<cudnnConvolutionFwdAlgo_t> {
...
+        conv.wdesc.desc,
+        wght,
+        conv.cdesc.desc,
+        conv.odesc.desc,
+        out,
+        1,
+        &algoCount,
+        &perfResults,
+        ws.data,
+        ws.size));
```
Comment (apaszke) [inline torch/csrc/cudnn/Conv.cpp]: I liked it more before 😞

File: torch/csrc/cudnn/Conv.cpp [hunk 3]
```diff
@@ -282,6 +409,11 @@ void findAlgorithm(
     return;
   }
 
+  if (deterministic) {
+    *algo = search::DEFAULT_ALGO;
```
Comment (apaszke) [inline torch/csrc/cudnn/Conv.cpp]: is there really only one deterministic algorithm in cuDNN?!

File: torch/csrc/cudnn/Conv.cpp [hunk 4]
```diff
@@ -402,9 +542,10 @@ static void check_bias_size(THVoidTensor* bias, THVoidTensor* weight, int groups
   }
 }
 
-static void check_expected_output_size_is_valid(THVoidTensor* input, THVoidTensor* output, THVoidTensor* weight,
-                                                const std::vector<int>& pad, const std::vector<int>& stride,
-                                                const std::vector<int>& dilation)
+static void check_expected_output_size_is_valid(
+    THVoidTensor* input, THVoidTensor* output, THVoidTensor* weight,
+    const std::vector<int>& pad, const std::vector<int>& stride,
+    const std::vector<int>& dilation)
```
Comment (apaszke) [inline torch/csrc/cudnn/Conv.cpp]: uhhh let's not change the formatting unless it's really necessary... I can't see what was wrong with these lines and it makes the diff soo much larger than it has to be (which means the review takes so much longer, or it's easy to skip important changes)

File: torch/csrc/autograd/functions/convolution.h [hunk 5]
```diff
@@ -29,6 +29,7 @@ struct ConvParams {
   std::vector<int> output_padding;
   int groups;
   bool benchmark;
+  bool deterministic;
```
Comment (apaszke) [inline torch/csrc/autograd/functions/convolution.h]: This field is never set anywhere, and since it's a primitive type, it is not initialized (which means it will randomly be true and false at different times).

File: torch/backends/cudnn/__init__.py [hunk 6]
```diff
@@ -58,6 +58,7 @@ def is_acceptable(tensor):
 
 _handles = {}
 
+deterministic = False
```
Comment (apaszke) [inline torch/backends/cudnn/__init__.py]: This flag is also unused.

**Comment by apaszke** [comment]
Please please please please don't change the style of the cpp files unless it's really needed for the PR. We don't have a strict style guide for C++ (our guideline is to just match the style of the file you're editing), and it creates a lot of unnecessary noise

File: torch/csrc/cudnn/Conv.cpp (hunk 2 above)
Comment (prigoyal) [reply torch/csrc/cudnn/Conv.cpp]: it was hard for me to parse the code so I made these changes

File: torch/csrc/cudnn/Conv.cpp (hunk 3 above)
Comment (prigoyal) [reply torch/csrc/cudnn/Conv.cpp]: that's what I saw in the cudnn documentation

**Comment by prigoyal** [comment]
thanks @soumith I clearly missed that. I thought there was a way to globally set the flags in __init__ and didn't look any further! I'll also look at the tests for this to make sure it works

File: torch/backends/cudnn/__init__.py (hunk 6 above)
Comment (prigoyal) [reply torch/backends/cudnn/__init__.py]: resolved in the latest update!

File: torch/csrc/autograd/functions/convolution.h (hunk 5 above)
Comment (prigoyal) [reply torch/csrc/autograd/functions/convolution.h]: resolved!

**Comment by prigoyal** [comment]
it seems that cudNN deterministic algo for example CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMPUTE_GEMM doesn't support dilated conv?  can someone confirm whether cudnn has deterministic dilated conv support? @ngimel

File: test/test_nn.py [hunk 7]
```diff
@@ -3054,8 +3072,8 @@ def test_conv_double_backward_cuda(self):
 
 class TestNNInit(TestCase):
     def setUp(self):
-        super(TestNNInit, self).setUp()
         random.seed(123)
+        torch.manual_seed(123)
```
Comment (prigoyal) [inline test/test_nn.py]: rebase and IDE cache issue, I'll revert this

**Comment by ngimel** [comment]
@prigoyal the way to find if there is a deterministic algo for a particular convolution is to run cudnnFind* and get all results (not just the first one). Starting from cudnn v6, the return structure of cudnnFind has a field that says if an algo is deterministic, so looping over returned algos one can find if there is a deterministic among them or not. 
It's true that CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMPUTE_GEMM does not support dilated conv, but there should be an algorithm that does. In fact, all forward algos are deterministic, so there should be no problems with forward. I still did not get an answer from cudnn if backward algorithms that support dilated conv are deterministic, but the return of cudnnFind should answer that question.

**Comment by ngimel** [comment]
Turns out there's no support for deterministic dilated gradient wrt data, and deterministic gradient wrt params was recently added, so it is probably only in cudnn v7, and not in v6.

**Comment by prigoyal** [comment]
thanks @ngimel for confirming :-)
in that case, it seems like the best way to proceed is to not have deterministic dilated conv yet in pytorch?

**Comment by apaszke** [comment]
We can just fall back to THCUNN if someone wants a deterministic dilated conv

**Comment by prigoyal** [comment]
@apaszke because cudnn doesn't have support for deterministic dilated gradient w.r.t. data, the dilated conv can't be truly deterministic. So like you suggested, we can have a fallback but I think we should also throw some message that user passed invalid arguments (i.e dilation > 1 and deterministic=True) ow they will wonder why their dilated conv is not giving deterministic result. In short, I think we should have a fallback and throw some warning to say something about this.

**Comment by apaszke** [comment]
IIRC THCUNN conv is just terribly slow, but is deterministic, so we can safely fall back to it. It would still be nice to raise an error and suggest turning deterministic off if it hurts your speed too much.

**Comment by soumith** [comment]
three changes required:

1. please use the determinism field as natalia suggested in https://github.com/pytorch/pytorch/pull/2893#issuecomment-333164213 instead of using DEFAULT_ALGO
2. Introduce a `bool forward` as a second argument to https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/convolution.cpp#L83. if is_dilated and backward mode and deterministic, then return use_cudnn to false.

**Comment by apaszke** [comment]
@soumith what's a `bool forward`?

**Comment by soumith** [comment]
as a second argument to the use_cudnn function. Natalia said there's only deterministic stuff missing for dilated + backward.

**Comment by soumith** [comment]
actually, probably simpler to just not introduce the `bool forward` argument to `use_cudnn` and disable cudnn in the case of `deterministic && dilated`

**Comment by apaszke** [comment]
Right. Also, we'll have to make sure THCUNN doesn't need any buffers from forward and we can safely fall back to it only in backward

**Comment by prigoyal** [comment]
@apaszke what do you mean by "we'll have to make sure THCUNN doesn't need any buffers from forward"? why would it require buffers from forward?

**Comment by soumith** [comment]
@prigoyal the THCUNN implementation might or might not expect that some Tensors that you give as input to the backward functions have been run through the forward function. if we do CuDNN forward and THCUNN backward, it wont hold true.

**Comment by prigoyal** [comment]
1) if dilation > 1 and determinisitc -> use_cudnn = False
2) if determinism and benchmark: we look over the perf results of all algorithms and return the best determinisitic algo. If there is a cudnn failure in this, we will return the default algorithm.
3) if determinism and not benchmark: return the default algorithm.

how do I check for whether THCUNN needs any buffers from forward? I am disabling cudnn for dilation & deterministic case only so far.

**Comment by soumith** [comment]
> how do I check for whether THCUNN needs any buffers from forward? I am disabling cudnn for dilation & deterministic case only so far.

You dont need to worry about this now with the current plan.

File: torch/csrc/cudnn/Conv.cpp [hunk 8]
```diff
@@ -142,17 +179,30 @@ size_t getMaxWorkspaceSize(cudnnHandle_t handle, const Convolution& conv, algo_t
     return max_ws_size;
 }
 
+template<typename perf_t>
+perf_t getBestAlgorithm(perf_t *perfResults, bool deterministic, int n_algo) {
+  if (deterministic) {
+    for (int i = 0; i < n_algo; i++) {
+      if (perfResults[i].status == CUDNN_STATUS_SUCCESS && perfResults[i].determinism == CUDNN_DETERMINISTIC) {
```
Comment (prigoyal) [inline torch/csrc/cudnn/Conv.cpp]: the algorithms are returned in sorted way by cudnn i.e. first is the fastest etc.
if there was error in benchmarking, we fallback to default algo which are also deterministic and could be a better choice to make

**Comment by prigoyal** [comment]
is this mergeable now? @apaszke @soumith

File: test/test_nn.py [hunk 9]
```diff
@@ -1770,6 +1770,24 @@ def test_Conv2d_inconsistent_types_on_GPU_with_cudnn(self):
         # but it should work with the same type
         nn.functional.conv2d(inputs.float(), weights.float(), bias.float())
 
+    @unittest.skipIf(not TEST_CUDA, 'CUDA not available')
+    @unittest.skipIf(not TEST_CUDNN, 'CUDNN not available')
+    def test_Conv2d_deterministic_on_GPU_with_cudnn(self):
+        inputs = Variable(torch.randn(4, 1, 7, 7).float().cuda())
+        weights = Variable(torch.randn(1, 1, 3, 3).double().cuda())
+        bias = Variable(torch.randn(1).double().cuda())
```
Comment (apaszke) [inline test/test_nn.py]: why are your weights of different type than input? how could that even pass a contbuild?

File: test/test_nn.py [hunk 10]
```diff
@@ -1770,6 +1770,24 @@ def test_Conv2d_inconsistent_types_on_GPU_with_cudnn(self):
...
+
+        torch.backends.cudnn.enabled = True
+        torch.backends.cudnn.benchmark = True
+        torch.backends.cudnn.deterministic = True
```
Comment (apaszke) [inline test/test_nn.py]: you have to implement a context manager that will restore these flags after the test. right now they will be just left like this forever, destroying the environment for other tests. It needs to be a context manager, so that it's executed even in case of an error.

**Review by apaszke — COMMENTED** [review]
C code looks good. Tests need to be fixed.

File: test/test_nn.py (hunk 9 above)
Comment (prigoyal) [reply test/test_nn.py]: I guess I took it from the previous tests and line 1787-1788 typecasts them before passing so contbuild should work.

File: test/test_nn.py (hunk 10 above)
Comment (prigoyal) [reply test/test_nn.py]: sure, that sounds like a good thing to do although I wondered why the context wasn't managed in rest of the tests...

Comment (apaszke) [reply test/test_nn.py]: it must have slipped through the cracks. We must have accidentally accepted such PRs, but it's breaking the tests

Comment (prigoyal) [reply test/test_nn.py]: agreed! I'll add the context here to this specific test for now and improve the other tests in separate PR

Comment (fmassa) [reply test/test_nn.py]: I think it would be useful to expose this context manager outside of the tests, so that in conjunction with the new `torch.random.fork_rng` from https://github.com/pytorch/pytorch/pull/2966 the users can now have an easy way to make their code fully-deterministic.

Comment (apaszke) [reply test/test_nn.py]: That's a really good idea. Let's do this!

**Comment by prigoyal** [comment]
added a `manage_context.py` file and hopefully this can be used in future for adding context managers for other things as well, right now, it only contains cudnn environment setting like enabled, benchmark, deterministic

File: torch/manage_context.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
+"""
+Manage various cudnn contexts
+"""
+
+import torch
+from contextlib import contextmanager
+
+
+@contextmanager
+def cudnn_context(use_cudnn=False, benchmark=False, deterministic=False, verbose=False):
```
Comment (apaszke) [inline torch/manage_context.py]: This should go to `torch.backends.cudnn` (somewhere close to where these flags are defined). I don't think we need a separate file for context managers. Let's keep them close to what they are needed for. I'd imagine it used like this:
```python
from torch.backends import cudnn

with cudnn.flags(enabled=True, benchmarks=False, deterministic=True):
    ...
```

Also, please use `enabled` instead of `use_cudnn`, so it matches the flag.

Comment (prigoyal) [reply torch/manage_context.py]: I was hoping that we could use this manage_context file as a general file to have context for other things as well as we come across use cases in future. I can move it to cudnn itself though if we don't see much utility for it yet

Comment (apaszke) [reply torch/manage_context.py]: We already have a few context manager, but we keep them in their respective modules. Let's keep this convention. I think it makes sense

Comment (prigoyal) [reply torch/manage_context.py]: done! this should be mergeable now after waiting for contbuild

File: torch/backends/cudnn/__init__.py [hunk 12]
```diff
@@ -81,6 +83,29 @@ def is_acceptable(tensor):
...
+    orig_enable = enabled
+    orig_benchmark = benchmark
+    orig_deterministic = deterministic
+    orig_verbose = verbose
+
+    # set the new values
+    enabled = enabled
+    benchmark = benchmark
+    deterministic = deterministic
+    verbose = verbose
```
Comment (apaszke) [inline torch/backends/cudnn/__init__.py]: um this doesn't work. The whole context manager is a no-op right now. You're just assigning local variable to the same variable, as the global flags got shadowed. You might need to fiddle with `globals()`, or use a helper function that will take arguments with different names (i.e. make `flags` call `_flags(_enabled=enabled, _benchmark=benchmark, ...)`

File: torch/backends/cudnn/__init__.py [hunk 13]
```diff
@@ -81,6 +83,29 @@ def is_acceptable(tensor):
 CUDNN_RNN_ALGO_PERSIST_DYNAMIC = 2
 
 
+@contextmanager
+def flags(enabled=False, benchmark=False, deterministic=False, verbose=False):
+    # capture the original values
+    orig_enable = enabled
+    orig_benchmark = benchmark
+    orig_deterministic = deterministic
+    orig_verbose = verbose
```
Comment (apaszke) [inline torch/backends/cudnn/__init__.py]: A shortcut for saving and restoring these flags would be this:
```python
orig_flags = enabled, benchmark, deterministic, verbose
try:
    ...
finally:
    enabled, benchmark, deterministic, verbose = orig_flags
```

File: torch/backends/cudnn/__init__.py (hunk 12 above)
Comment (prigoyal) [reply torch/backends/cudnn/__init__.py]: yes, n00b errors, my bad! thanks for catching!

**Comment by prigoyal** [comment]
ping! @apaszke :)

File: test/test_nn.py [hunk 14]
```diff
@@ -1770,6 +1760,17 @@ def test_Conv2d_inconsistent_types_on_GPU_with_cudnn(self):
...
+    @unittest.skipIf(not TEST_CUDA, 'CUDA not available')
+    @unittest.skipIf(not TEST_CUDNN, 'CUDNN not available')
+    def test_Conv2d_deterministic_cudnn(self):
+        inputs = Variable(torch.randn(4, 1, 7, 7).float().cuda())
+        weights = Variable(torch.randn(1, 1, 3, 3).float().cuda())
+        bias = Variable(torch.randn(1).float().cuda())
+        with cudnn.flags(enabled=True, benchmark=True, deterministic=True):
+            f1 = nn.functional.conv2d(inputs, weights, bias)
+            f2 = nn.functional.conv2d(inputs, weights, bias)
+            self.assertEqual(f1, f2)
```
Comment (apaszke) [inline test/test_nn.py]: The important part of deterministic is also backward part (e.g. for MaxPool). We should check that the grads are equal.

Additionally, non-determinism is often visible at the very small scale (think 1e-6), and `self.assertEqual` compares the tensors pointwise and accepts them as long as no element differs by more than 1e-4. Add a third `0` argument, to force bitwise equality.

File: torch/backends/cudnn/__init__.py [hunk 15]
```diff
@@ -81,6 +83,23 @@ def is_acceptable(tensor):
...
+
+
+@contextmanager
+def flags(enabled=False, benchmark=False, deterministic=False, verbose=False):
+    orig_flags = set_flags(enabled, benchmark, deterministic, verbose)
+    try:
+        yield
+    finally:
+        # recover the previous values
+        set_flags(orig_flags[0], orig_flags[1], orig_flags[2], orig_flags[3])
```
Comment (apaszke) [inline torch/backends/cudnn/__init__.py]: nit: you can do `*orig_flags` here

File: test/test_nn.py (hunk 14 above)
Comment (prigoyal) [reply test/test_nn.py]: This particular PR is about conv and not the MaxPool layer
`self.assertEqual` compares tensors and accepts then for 1e-5 but I'll add the third argument 0 as well for the bitwise equality

Comment (apaszke) [reply test/test_nn.py]: I know, I just don't know the places in conv algos that can cause them to be nondeterministic, while I know the parts that can make MaxPool backward non-deterministic. That's why I used it as an example.

**Comment by prigoyal** [comment]
is this mergeable now?
